rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,242 @@
+
+ - Changes in dependencies can require dependents to add an override, causing packages to be moved back and forth between unit directories and the general `pkgs` tree, worsening the problem as people have to decide categories *again*.
+
+ - When lifting the restriction, the reviewers have to adapt, again leading to unnecessary requests for changes in pull requests.
+ 
+ - We'd be protracting the migration by unnecessary gatekeeping or discovering some problem late.
+
+That said, we did identify risks:
+
+ - We might get something wrong, and while we plan to incrementally migrate Nixpkgs to this new system anyway, starting with fewer units is good.
+    - Mitigation: only automate the renames of simple (`callPackage path { }`) calls, to keep the initial change small
+ 
+ - We might not focus enough on the foundation, while we could more easily relax requirements later.
+    - After more discussion, we feel confident that the manual `callPackage` calls are unlikely to cause issues that we wouldn't otherwise have.
+
+# Recommend a `callPackage` pattern with default arguments
+
+> - While this RFC doesn't address expressions where the second `callPackage` argument isn't `{}`, there is an easy way to transition to an argument of `{}`: For every attribute of the form `name = attrs.value;` in the argument, make sure `attrs` is in the arguments of the file, then add `name ? attrs.value` to the arguments. Then the expression in `all-packages.nix` can too be auto-called
+>   - Don't do this for `name = value` pairs though, that's an alias-like thing
+
+`callPackage` does not favor the default argument when both a default argument and a value in `pkgs` exist. Changing the semantics of `callPackage` is out of scope.
+
+# Allow `callPackage` arguments to be specified in `<unit>/args.nix`
+
+The idea was to expand the auto-calling logic according to:
+
+Unit directories are automatically discovered and transformed to a definition of the form
+```
+# If args.nix doesn't exist
+pkgs.${name} = pkgs.callPackage ${unitDir}/pkg-fun.nix {}
+# If args.nix does exists
+pkgs.${name} = pkgs.callPackage ${unitDir}/pkg-fun.nix (import ${unitDir}/args.nix pkgs);
+```
+
+Pro:
+ - It makes another class of packages uniform, by picking a solution with restricted expressive power.
+
+Con:
+ - It does not solve the contributor experience problem of having to many rules.
+ - `args.nix` is another pattern that contributors need to learn how to use, as we have seen that it is not immediately obvious to everyone how it works.
+ - A CI check can mitigate the possible lack of uniformity, and we see a simple implementation strategy for it.
+ - This keeps the contents of the unit directories simple and a bit more uniform than with optional `args.nix` files.
```

Comment:
Does this solve that the arguments are not well typed and not machine discoverable? If we go this route, we should try to address this.

Reply:
No it wouldn't fix that problem. This is definitely something for the Nixpkgs Architecture Team though. Definitely out of scope for this RFC though.

Reply:
Types without Nix-side support would be a large performance hit, no? 

I wonder how much discoverability of arguments we can get even with the current structure…